gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcEventRepository.java

```diff
@@ -280,6 +280,47 @@ public List<Event> search(EventCriteria filter) {
         return queryEvents(builder.toString(), args);
     }
 
+    @Override
+    public Event createOrUpdateHeartbeat(Event event) throws TechnicalException {
+        if (event == null || event.getId() == null) {
+            throw new IllegalStateException("Event to create or update must have an id");
+        }
+
+        if (existsById(event.getId())) {
+            final int updatedEventCount = updateHeartbeatEventDate(event);
+            int updatedEventPropertyCount = -1;
+            // update properties only if event was correctly updated
+            if (updatedEventCount == 1) {
+                updatedEventPropertyCount = updateHeartbeatEventProperties(event);
+            }
+
+            return updatedEventPropertyCount == 1 ? event : null;
+        }
+        return create(event);
+    }
+
+    private int updateHeartbeatEventDate(Event event) {
+        return jdbcTemplate.update("update " + this.tableName + " set updated_at = ? where id = ?", event.getUpdatedAt(), event.getId());
+    }
+
+    private int updateHeartbeatEventProperties(Event event) {
+        return jdbcTemplate.update(
+            "update " + EVENT_PROPERTIES + " set property_value = ? where event_id = ? and property_key = 'last_heartbeat_at'",
```

Comment:
As a performance improvment, is it relevant to add an index on event_id + property_key, according to the new query here ?
I see the index `idx_${gravitee_prefix}eventproperties_propertykey` for property_key, but it doesn't include id.

Reply:
discussed